test-framework/core/src/main/java/org/keycloak/testframework/realm/UserSupplier.java

```diff
@@ -41,12 +49,41 @@ public ManagedUser getValue(InstanceContext<ManagedUser, InjectUser> instanceCon
             if (Status.CONFLICT.equals(Status.fromStatusCode(response.getStatus()))) {
                 throw new IllegalStateException("User already exist with username: " + userRepresentation.getUsername());
             }
-            String uuid = ApiUtil.handleCreatedResponse(response);
+            String userUuid = ApiUtil.handleCreatedResponse(response);
+
+            instanceContext.addNote(USER_UUID_KEY, userUuid);
 
-            instanceContext.addNote(USER_UUID_KEY, uuid);
+            UserResource userResource = realm.admin().users().get(userUuid);
+            userRepresentation.setId(userUuid);
 
-            UserResource userResource = realm.admin().users().get(uuid);
-            userRepresentation.setId(uuid);
+            // let's assign the client roles with as much optimization as possible
+            Map<String, List<String>> clientRoles = userRepresentation.getClientRoles();
+            if (clientRoles == null || clientRoles.isEmpty()) {
+                return new ManagedUser(userRepresentation, userResource);
+            }
+            // replace map keys (instead of the clientId we need the client's uuid)
+            ClientsResource clientsResource = realm.admin().clients();
+            Map<String, List<String>> clientRolesUuid = clientRoles.entrySet().stream().collect(Collectors.toMap(
+                    entry -> clientsResource.findByClientId(entry.getKey()).stream()
+                            .findFirst()
+                            .orElseThrow()
+                            .getId(),
+                    Map.Entry::getValue
+            ));
+            // for each client uuid, get the RoleScopeResource
+            RoleMappingResource roleMappingResource = userResource.roles();
+            Map<String, RoleScopeResource> roleScopeResources = clientRolesUuid.keySet().stream().collect(Collectors.toMap(
+                    key -> key,
+                    roleMappingResource::clientLevel
+            ));
+            // replace the string described roles with actual role representations
+            Map<String, List<RoleRepresentation>> clientRolesRepresentations = clientRolesUuid.entrySet().stream().collect(Collectors.toMap(
+                    Map.Entry::getKey,
+                    entry -> roleScopeResources.get(entry.getKey()).listAvailable().stream()
+                            .filter(role -> entry.getValue().contains(role.getName())).toList()
+            ));
+            // assign user the client roles
+            clientRolesRepresentations.forEach((key, value) -> roleScopeResources.get(key).add(value));
```

Comment:
I tried optimizing admin API calls, but the code readability worsened. WDYT? @stianst @mabartos

Reply:
Here's the last "optimization" change: https://github.com/keycloak/keycloak/pull/36828/commits/ad98234cc9352daffe9c8d0678e5e442a4ec1b63?diff=split&w=0